File: frontend/src/types/analytics.ts
/**
 * Analytics API Types
 *
 * Type definitions matching backend Pydantic models.
 * All field names use snake_case to match API response.
 */

/**
 * Member item for analytics selector dropdown
 */
export interface MemberListItem {
  readonly id: string
  readonly name: string
  readonly is_active: boolean
  readonly contribution_rank: number | null
  readonly group: string | null
}

/**
 * Single period data point for member trend chart
 */
export interface MemberTrendItem {
  // Period identification
  readonly period_id: string
  readonly period_number: number
  readonly period_label: string
  readonly start_date: string // ISO date string
  readonly end_date: string // ISO date string
  readonly days: number

  // Daily averages
  readonly daily_contribution: number
  readonly daily_merit: number
  readonly daily_assist: number
  readonly daily_donation: number

  // Period totals (diff values)
  readonly contribution_diff: number
  readonly merit_diff: number
  readonly assist_diff: number
  readonly donation_diff: number
  readonly power_diff: number

  // Rank info
  readonly start_rank: number | null
  readonly end_rank: number
  readonly rank_change: number | null

  // State info
  readonly end_power: number
  readonly end_state: string | null
  readonly end_group: string | null
  readonly is_new_member: boolean

  // Alliance averages for comparison
  readonly alliance_avg_contribution: number
  readonly alliance_avg_merit: number
  readonly alliance_avg_assist: number
  readonly alliance_avg_donation: number
  readonly alliance_avg_power: number
  readonly alliance_member_count: number

  // Alliance medians for comparison
  readonly alliance_median_contribution: number
  readonly alliance_median_merit: number
  readonly alliance_median_assist: number
  readonly alliance_median_donation: number
  readonly alliance_median_power: number
}

/**
 * Season-to-date summary for a member
 */
export interface SeasonSummaryResponse {
  // Period counts
  readonly period_count: number
  readonly total_days: number

  // Season totals
  readonly total_contribution: number
  readonly total_merit: number
  readonly total_assist: number
  readonly total_donation: number
  readonly total_power_change: number

  // Season daily averages
  readonly avg_daily_contribution: number
  readonly avg_daily_merit: number
  readonly avg_daily_assist: number
  readonly avg_daily_donation: number
  readonly avg_power: number

  // Current status
  readonly current_rank: number
  readonly rank_change_season: number | null
  readonly current_power: number
  readonly current_group: string | null
  readonly current_state: string | null
}

/**
 * Alliance-wide average and median metrics for a period
 */
export interface AllianceAveragesResponse {
  readonly member_count: number
  // Averages
  readonly avg_daily_contribution: number
  readonly avg_daily_merit: number
  readonly avg_daily_assist: number
  readonly avg_daily_donation: number
  readonly avg_power: number
  // Medians
  readonly median_daily_contribution: number
  readonly median_daily_merit: number
  readonly median_daily_assist: number
  readonly median_daily_donation: number
  readonly median_power: number
}

/**
 * Alliance averages for a single period in trend
 */
export interface AllianceTrendItem {
  readonly period_id: string
  readonly period_number: number
  readonly period_label: string
  readonly member_count: number
  readonly avg_daily_contribution: number
  readonly avg_daily_merit: number
  readonly avg_daily_assist: number
  readonly avg_daily_donation: number
}

// =============================================================================
// Group Analytics Types
// =============================================================================

/**
 * Group item for group selector dropdown
 */
export interface GroupListItem {
  readonly name: string
  readonly member_count: number
}

/**
 * Group statistics based on latest period data
 */
export interface GroupStats {
  readonly group_name: string
  readonly member_count: number

  // Person-day averages (人日均)
  readonly avg_daily_contribution: number
  readonly avg_daily_merit: number
  readonly avg_daily_assist: number
  readonly avg_daily_donation: number
  readonly avg_power: number

  // Rank statistics
  readonly avg_rank: number
  readonly best_rank: number
  readonly worst_rank: number

  // Contribution distribution (box plot data)
  readonly contribution_min: number
  readonly contribution_q1: number
  readonly contribution_median: number
  readonly contribution_q3: number
  readonly contribution_max: number
  readonly contribution_cv: number

  // Merit distribution (box plot data)
  readonly merit_min: number
  readonly merit_q1: number
  readonly merit_median: number
  readonly merit_q3: number
  readonly merit_max: number
  readonly merit_cv: number
}

/**
 * Member within a group with performance metrics
 */
export interface GroupMember {
  readonly id: string
  readonly name: string
  readonly contribution_rank: number
  readonly daily_contribution: number
  readonly daily_merit: number
  readonly daily_assist: number
  readonly daily_donation: number
  readonly power: number
  readonly rank_change: number | null
  readonly contribution_change: number | null
  readonly merit_change: number | null
}

/**
 * Group performance for a single period
 */
export interface GroupTrendItem {
  readonly period_label: string
  readonly period_number: number
  readonly start_date: string // ISO date string
  readonly end_date: string // ISO date string
  readonly days: number
  readonly avg_rank: number
  readonly avg_contribution: number
  readonly avg_merit: number
  readonly avg_assist: number
  readonly avg_donation: number
  readonly avg_power: number
  readonly member_count: number
}

/**
 * Complete group analytics data (aggregated response)
 */
export interface GroupAnalyticsResponse {
  readonly stats: GroupStats
  readonly members: readonly GroupMember[]
  readonly trends: readonly GroupTrendItem[]
  readonly alliance_averages: AllianceAveragesResponse
}

/**
 * Group summary for comparison across all groups
 */
export interface GroupComparisonItem {
  readonly name: string
  readonly avg_daily_merit: number
  readonly avg_rank: number
  readonly member_count: number
}

// =============================================================================
// Alliance Analytics Types
// =============================================================================

/**
 * Alliance-wide metrics summary for KPI cards
 */
export interface AllianceSummary {
  readonly member_count: number
  readonly avg_daily_contribution: number
  readonly avg_daily_merit: number
  readonly avg_daily_assist: number
  readonly avg_daily_donation: number
  readonly avg_power: number
  readonly median_daily_contribution: number
  readonly median_daily_merit: number
  readonly contribution_change_pct: number | null
  readonly merit_change_pct: number | null
  readonly power_change_pct: number | null
}

/**
 * Enhanced trend item with median values for charts
 */
export interface AllianceTrendWithMedian {
  readonly period_id: string
  readonly period_number: number
  readonly period_label: string
  readonly start_date: string
  readonly end_date: string
  readonly days: number
  readonly member_count: number
  // Averages
  readonly avg_daily_contribution: number
  readonly avg_daily_merit: number
  readonly avg_daily_assist: number
  readonly avg_daily_donation: number
  readonly avg_power: number
  // Medians
  readonly median_daily_contribution: number
  readonly median_daily_merit: number
  readonly median_daily_assist: number
  readonly median_daily_donation: number
}

/**
 * Histogram bin for distribution charts
 */
export interface DistributionBin {
  readonly range: string
  readonly min_value: number
  readonly max_value: number
  readonly count: number
}

/**
 * Distribution histograms for contribution and merit
 */
export interface DistributionData {
  readonly contribution: readonly DistributionBin[]
  readonly merit: readonly DistributionBin[]
}

/**
 * Group stats with box plot data for alliance analytics
 */
export interface GroupStatsWithBoxPlot {
  readonly name: string
  readonly member_count: number
  readonly avg_daily_contribution: number
  readonly avg_daily_merit: number
  readonly avg_rank: number
  readonly avg_power: number
  readonly contribution_cv: number
  // Contribution box plot
  readonly contribution_min: number
  readonly contribution_q1: number
  readonly contribution_median: number
  readonly contribution_q3: number
  readonly contribution_max: number
  // Merit box plot
  readonly merit_min: number
  readonly merit_q1: number
  readonly merit_median: number
  readonly merit_q3: number
  readonly merit_max: number
}

/**
 * Top/Bottom performer member item
 */
export interface PerformerItem {
  readonly member_id: string
  readonly name: string
  readonly group: string | null
  readonly daily_contribution: number
  readonly daily_merit: number
  readonly daily_assist: number
  readonly rank: number
  readonly rank_change: number | null
  readonly merit_change: number | null
  readonly assist_change: number | null
}

/**
 * Member needing attention
 */
export interface AttentionItem {
  readonly member_id: string
  readonly name: string
  readonly group: string | null
  readonly daily_contribution: number
  readonly rank: number
  readonly rank_change: number | null
  readonly reason: string
}

/**
 * Current period metadata
 */
export interface PeriodInfo {
  readonly period_id: string
  readonly period_number: number
  readonly period_label: string
  readonly start_date: string
  readonly end_date: string
  readonly days: number
}

/**
 * Complete alliance analytics data for AllianceAnalytics page
 */
export interface AllianceAnalyticsResponse {
  readonly summary: AllianceSummary
  readonly trends: readonly AllianceTrendWithMedian[]
  readonly distributions: DistributionData
  readonly groups: readonly GroupStatsWithBoxPlot[]
  readonly top_performers: readonly PerformerItem[]
  readonly bottom_performers: readonly PerformerItem[]
  readonly needs_attention: readonly AttentionItem[]
  readonly current_period: PeriodInfo
}

File: backend/src/api/v1/schemas/analytics.py
"""
Analytics API Schemas

Pydantic models for analytics API request/response validation.

Follows CLAUDE.md:
- Pydantic V2 syntax with ConfigDict
- snake_case field names
- Explicit type hints
"""

from datetime import date

from pydantic import BaseModel, ConfigDict, Field


class MemberListItem(BaseModel):
    """Member item for analytics selector dropdown"""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., description="Member UUID as string")
    name: str = Field(..., description="Member display name")
    is_active: bool = Field(..., description="Whether member is currently active")
    contribution_rank: int | None = Field(None, description="Latest contribution rank")
    group: str | None = Field(None, description="Current group assignment")


class MemberTrendItem(BaseModel):
    """Single period data point for member trend chart"""

    model_config = ConfigDict(from_attributes=True)

    # Period identification
    period_id: str = Field(..., description="Period UUID as string")
    period_number: int = Field(..., ge=1, description="Period number within season")
    period_label: str = Field(..., description="Display label (e.g., '10/02-10/09')")
    start_date: date = Field(..., description="Period start date")
    end_date: date = Field(..., description="Period end date")
    days: int = Field(..., ge=1, description="Number of days in period")

    # Daily averages
    daily_contribution: float = Field(..., ge=0, description="Daily average contribution")
    daily_merit: float = Field(..., ge=0, description="Daily average merit")
    daily_assist: float = Field(..., ge=0, description="Daily average assist")
    daily_donation: float = Field(..., ge=0, description="Daily average donation")

    # Period totals (diff values)
    contribution_diff: int = Field(..., description="Total contribution gained in period")
    merit_diff: int = Field(..., description="Total merit gained in period")
    assist_diff: int = Field(..., description="Total assists in period")
    donation_diff: int = Field(..., description="Total donation in period")
    power_diff: int = Field(..., description="Power change in period (can be negative)")

    # Rank info
    start_rank: int | None = Field(None, description="Rank at period start (None for new members)")
    end_rank: int = Field(..., description="Rank at period end")
    rank_change: int | None = Field(None, description="Rank change (positive = improved)")

    # State info
    end_power: int = Field(..., description="Power at period end")
    end_state: str | None = Field(None, description="State/region at period end")
    end_group: str | None = Field(None, description="Group assignment at period end")
    is_new_member: bool = Field(..., description="Whether this is member's first period")

    # Alliance averages for comparison
    alliance_avg_contribution: float = Field(0, description="Alliance avg daily contribution")
    alliance_avg_merit: float = Field(0, description="Alliance avg daily merit")
    alliance_avg_assist: float = Field(0, description="Alliance avg daily assist")
    alliance_avg_donation: float = Field(0, description="Alliance avg daily donation")
    alliance_avg_power: float = Field(0, description="Alliance avg power")
    alliance_member_count: int = Field(0, description="Total members in alliance for this period")

    # Alliance medians for comparison
    alliance_median_contribution: float = Field(0, description="Alliance median daily contribution")
    alliance_median_merit: float = Field(0, description="Alliance median daily merit")
    alliance_median_assist: float = Field(0, description="Alliance median daily assist")
    alliance_median_donation: float = Field(0, description="Alliance median daily donation")
    alliance_median_power: float = Field(0, description="Alliance median power")


class SeasonSummaryResponse(BaseModel):
    """Season-to-date summary for a member"""

    model_config = ConfigDict(from_attributes=True)

    # Period counts
    period_count: int = Field(..., ge=0, description="Number of periods included")
    total_days: int = Field(..., ge=0, description="Total days across all periods")

    # Season totals
    total_contribution: int = Field(..., description="Total contribution across season")
    total_merit: int = Field(..., description="Total merit across season")
    total_assist: int = Field(..., description="Total assists across season")
    total_donation: int = Field(..., description="Total donation across season")
    total_power_change: int = Field(..., description="Net power change across season")

    # Season daily averages
    avg_daily_contribution: float = Field(..., ge=0, description="Average daily contribution")
    avg_daily_merit: float = Field(..., ge=0, description="Average daily merit")
    avg_daily_assist: float = Field(..., ge=0, description="Average daily assist")
    avg_daily_donation: float = Field(..., ge=0, description="Average daily donation")
    avg_power: float = Field(..., ge=0, description="Average power across all periods")

    # Current status
    current_rank: int = Field(..., description="Current rank (from latest period)")
    rank_change_season: int | None = Field(None, description="Rank change since season start")
    current_power: int = Field(..., description="Current power (from latest period)")
    current_group: str | None = Field(None, description="Current group assignment")
    current_state: str | None = Field(None, description="Current state/region")


class AllianceAveragesResponse(BaseModel):
    """Alliance-wide average and median metrics for a period"""

    model_config = ConfigDict(from_attributes=True)

    member_count: int = Field(..., ge=0, description="Number of members included")
    # Averages
    avg_daily_contribution: float = Field(..., ge=0, description="Average daily contribution")
    avg_daily_merit: float = Field(..., ge=0, description="Average daily merit")
    avg_daily_assist: float = Field(..., ge=0, description="Average daily assist")
    avg_daily_donation: float = Field(..., ge=0, description="Average daily donation")
    avg_power: float = Field(0, ge=0, description="Average power (latest snapshot)")
    # Medians
    median_daily_contribution: float = Field(0, ge=0, description="Median daily contribution")
    median_daily_merit: float = Field(0, ge=0, description="Median daily merit")
    median_daily_assist: float = Field(0, ge=0, description="Median daily assist")
    median_daily_donation: float = Field(0, ge=0, description="Median daily donation")
    median_power: float = Field(0, ge=0, description="Median power (latest snapshot)")


class AllianceTrendItem(BaseModel):
    """Alliance averages for a single period"""

    model_config = ConfigDict(from_attributes=True)

    period_id: str = Field(..., description="Period UUID as string")
    period_number: int = Field(..., ge=1, description="Period number within season")
    period_label: str = Field(..., description="Display label")
    member_count: int = Field(..., ge=0, description="Number of members")
    avg_daily_contribution: float = Field(..., ge=0)
    avg_daily_merit: float = Field(..., ge=0)
    avg_daily_assist: float = Field(..., ge=0)
    avg_daily_donation: float = Field(..., ge=0)


class MemberMetricsSnapshot(BaseModel):
    """Member metrics for a single period comparison"""

    model_config = ConfigDict(from_attributes=True)

    daily_contribution: float = Field(..., ge=0)
    daily_merit: float = Field(..., ge=0)
    daily_assist: float = Field(..., ge=0)
    daily_donation: float = Field(..., ge=0)
    end_rank: int = Field(..., ge=1)
    rank_change: int | None = Field(None)
    end_power: int = Field(..., ge=0)
    power_diff: int = Field(...)
    is_new_member: bool = Field(...)


class AllianceMetricsAverage(BaseModel):
    """Alliance average metrics for comparison"""

    model_config = ConfigDict(from_attributes=True)

    daily_contribution: float = Field(..., ge=0)
    daily_merit: float = Field(..., ge=0)
    daily_assist: float = Field(..., ge=0)
    daily_donation: float = Field(..., ge=0)


class AllianceMetricsMedian(BaseModel):
    """Alliance median metrics for comparison"""

    model_config = ConfigDict(from_attributes=True)

    daily_contribution: float = Field(..., ge=0)
    daily_merit: float = Field(..., ge=0)
    daily_assist: float = Field(..., ge=0)
    daily_donation: float = Field(..., ge=0)


class MemberComparisonResponse(BaseModel):
    """Member metrics with alliance averages and medians for comparison"""

    model_config = ConfigDict(from_attributes=True)

    member: MemberMetricsSnapshot = Field(..., description="Member's metrics")
    alliance_avg: AllianceMetricsAverage = Field(..., description="Alliance averages")
    alliance_median: AllianceMetricsMedian = Field(..., description="Alliance medians")
    total_members: int = Field(..., ge=0, description="Total members in comparison")


# ============================================================================
# Group Analytics Schemas
# ============================================================================


class GroupListItem(BaseModel):
    """Group item for group selector dropdown"""

    model_config = ConfigDict(from_attributes=True)

    name: str = Field(..., description="Group name")
    member_count: int = Field(..., ge=0, description="Number of members in group")


class GroupStats(BaseModel):
    """Group statistics based on latest period data"""

    model_config = ConfigDict(from_attributes=True)

    group_name: str = Field(..., description="Group name")
    member_count: int = Field(..., ge=0, description="Number of members")

    # Person-day averages (人日均 - core comparison metrics)
    avg_daily_contribution: float = Field(..., ge=0, description="Average daily contribution per member")
    avg_daily_merit: float = Field(..., ge=0, description="Average daily merit per member")
    avg_daily_assist: float = Field(..., ge=0, description="Average daily assist per member")
    avg_daily_donation: float = Field(..., ge=0, description="Average daily donation per member")
    avg_power: float = Field(..., ge=0, description="Average power per member")

    # Rank statistics
    avg_rank: float = Field(..., description="Average contribution rank")
    best_rank: int = Field(..., ge=1, description="Best (lowest) rank in group")
    worst_rank: int = Field(..., ge=1, description="Worst (highest) rank in group")

    # Contribution distribution (box plot data)
    contribution_min: float = Field(..., ge=0, description="Minimum daily contribution")
    contribution_q1: float = Field(..., ge=0, description="Contribution first quartile (25th percentile)")
    contribution_median: float = Field(..., ge=0, description="Contribution median (50th percentile)")
    contribution_q3: float = Field(..., ge=0, description="Contribution third quartile (75th percentile)")
    contribution_max: float = Field(..., ge=0, description="Maximum daily contribution")
    contribution_cv: float = Field(..., ge=0, description="Contribution coefficient of variation (std/mean)")

    # Merit distribution (box plot data)
    merit_min: float = Field(..., ge=0, description="Minimum daily merit")
    merit_q1: float = Field(..., ge=0, description="First quartile (25th percentile)")
    merit_median: float = Field(..., ge=0, description="Median (50th percentile)")
    merit_q3: float = Field(..., ge=0, description="Third quartile (75th percentile)")
    merit_max: float = Field(..., ge=0, description="Maximum daily merit")
    merit_cv: float = Field(..., ge=0, description="Coefficient of variation (std/mean)")


class GroupMember(BaseModel):
    """Member within a group with performance metrics"""

    model_config = ConfigDict(from_attributes=True)

    id: str = Field(..., description="Member UUID as string")
    name: str = Field(..., description="Member display name")
    contribution_rank: int = Field(..., ge=1, description="Current contribution rank")
    daily_contribution: float = Field(..., ge=0, description="Daily average contribution")
    daily_merit: float = Field(..., ge=0, description="Daily average merit")
    daily_assist: float = Field(..., ge=0, description="Daily average assist")
    daily_donation: float = Field(..., ge=0, description="Daily average donation")
    power: int = Field(..., ge=0, description="Current power value")
    rank_change: int | None = Field(None, description="Rank change from previous period")
    contribution_change: float | None = Field(None, description="Daily contribution change from previous period")
    merit_change: float | None = Field(None, description="Daily merit change from previous period")


class GroupTrendItem(BaseModel):
    """Group performance for a single period"""

    model_config = ConfigDict(from_attributes=True)

    period_label: str = Field(..., description="Display label (e.g., '10/02-10/09')")
    period_number: int = Field(..., ge=1, description="Period number within season")
    start_date: date = Field(..., description="Period start date")
    end_date: date = Field(..., description="Period end date")
    days: int = Field(..., ge=1, description="Number of days in period")
    avg_rank: float = Field(..., description="Average contribution rank for period")
    avg_contribution: float = Field(..., ge=0, description="Average daily contribution for period")
    avg_merit: float = Field(..., ge=0, description="Average daily merit for period")
    avg_assist: float = Field(..., ge=0, description="Average daily assist for period")
    avg_donation: float = Field(..., ge=0, description="Average daily donation for period")
    avg_power: float = Field(..., ge=0, description="Average power for period")
    member_count: int = Field(..., ge=0, description="Number of members in group for period")


class GroupAnalyticsResponse(BaseModel):
    """Complete group analytics data (aggregated response)"""

    model_config = ConfigDict(from_attributes=True)

    stats: GroupStats = Field(..., description="Group statistics from latest period")
    members: list[GroupMember] = Field(..., description="Members in the group")
    trends: list[GroupTrendItem] = Field(..., description="Performance trend by period")
    alliance_averages: AllianceAveragesResponse = Field(
        ..., description="Alliance averages for comparison baseline"
    )


class GroupComparisonItem(BaseModel):
    """Group summary for comparison across all groups"""

    model_config = ConfigDict(from_attributes=True)

    name: str = Field(..., description="Group name")
    avg_daily_merit: float = Field(..., ge=0, description="Average daily merit")
    avg_rank: float = Field(..., description="Average contribution rank")
    member_count: int = Field(..., ge=0, description="Number of members")


# ============================================================================
# Alliance Analytics Schemas
# ============================================================================


class AllianceSummary(BaseModel):
    """Alliance-wide metrics summary for KPI cards"""

    model_config = ConfigDict(from_attributes=True)

    member_count: int = Field(..., ge=0, description="Number of active members")
    avg_daily_contribution: float = Field(..., ge=0)
    avg_daily_merit: float = Field(..., ge=0)
    avg_daily_assist: float = Field(..., ge=0)
    avg_daily_donation: float = Field(..., ge=0)
    avg_power: float = Field(..., ge=0)
    median_daily_contribution: float = Field(..., ge=0)
    median_daily_merit: float = Field(..., ge=0)
    # Change percentages vs previous period (None for season view)
    contribution_change_pct: float | None = Field(None, description="Contribution change % vs previous period")
    merit_change_pct: float | None = Field(None, description="Merit change % vs previous period")
    power_change_pct: float | None = Field(None, description="Power change % vs previous period")


class AllianceTrendWithMedian(BaseModel):
    """Enhanced trend item with median values for charts"""

    model_config = ConfigDict(from_attributes=True)

    period_id: str = Field(..., description="Period UUID as string")
    period_number: int = Field(..., ge=1)
    period_label: str = Field(..., description="Display label (e.g., '10/02-10/09')")
    start_date: str = Field(..., description="ISO date string")
    end_date: str = Field(..., description="ISO date string")
    days: int = Field(..., ge=1)
    member_count: int = Field(..., ge=0)
    # Averages
    avg_daily_contribution: float = Field(..., ge=0)
    avg_daily_merit: float = Field(..., ge=0)
    avg_daily_assist: float = Field(..., ge=0)
    avg_daily_donation: float = Field(..., ge=0)
    avg_power: float = Field(..., ge=0)
    # Medians
    median_daily_contribution: float = Field(..., ge=0)
    median_daily_merit: float = Field(..., ge=0)
    median_daily_assist: float = Field(..., ge=0)
    median_daily_donation: float = Field(..., ge=0)


class DistributionBin(BaseModel):
    """Histogram bin for distribution charts"""

    model_config = ConfigDict(from_attributes=True)

    range: str = Field(..., description="Display range label (e.g., '0-5K', '5K-10K')")
    min_value: float = Field(..., ge=0, description="Bin minimum value (inclusive)")
    max_value: float = Field(..., description="Bin maximum value (exclusive, except last bin)")
    count: int = Field(..., ge=0, description="Number of members in this bin")


class DistributionData(BaseModel):
    """Distribution histograms for contribution and merit"""

    model_config = ConfigDict(from_attributes=True)

    contribution: list[DistributionBin] = Field(..., description="Contribution distribution bins")
    merit: list[DistributionBin] = Field(..., description="Merit distribution bins")


class GroupStatsWithBoxPlot(BaseModel):
    """Group stats with box plot data for alliance analytics"""

    model_config = ConfigDict(from_attributes=True)

    name: str = Field(..., description="Group name")
    member_count: int = Field(..., ge=0)
    avg_daily_contribution: float = Field(..., ge=0)
    avg_daily_merit: float = Field(..., ge=0)
    avg_rank: float = Field(...)
    avg_power: float = Field(..., ge=0)
    contribution_cv: float = Field(..., ge=0, description="Coefficient of variation")
    # Contribution box plot
    contribution_min: float = Field(..., ge=0)
    contribution_q1: float = Field(..., ge=0)
    contribution_median: float = Field(..., ge=0)
    contribution_q3: float = Field(..., ge=0)
    contribution_max: float = Field(..., ge=0)
    # Merit box plot
    merit_min: float = Field(..., ge=0)
    merit_q1: float = Field(..., ge=0)
    merit_median: float = Field(..., ge=0)
    merit_q3: float = Field(..., ge=0)
    merit_max: float = Field(..., ge=0)


class PerformerItem(BaseModel):
    """Top/Bottom performer member item"""

    model_config = ConfigDict(from_attributes=True)

    member_id: str = Field(..., description="Member UUID as string")
    name: str = Field(..., description="Member display name")
    group: str | None = Field(None, description="Group assignment")
    daily_contribution: float = Field(..., ge=0)
    daily_merit: float = Field(..., ge=0)
    daily_assist: float = Field(..., ge=0)
    rank: int = Field(..., ge=1)
    rank_change: int | None = Field(None, description="Rank change (positive = improved)")
    merit_change: float | None = Field(None, description="Daily merit change vs previous period")
    assist_change: float | None = Field(None, description="Daily assist change vs previous period")


class AttentionItem(BaseModel):
    """Member needing attention"""

    model_config = ConfigDict(from_attributes=True)

    member_id: str = Field(..., description="Member UUID as string")
    name: str = Field(..., description="Member display name")
    group: str | None = Field(None, description="Group assignment")
    daily_contribution: float = Field(..., ge=0)
    rank: int = Field(..., ge=1)
    rank_change: int | None = Field(None)
    reason: str = Field(..., description="Reason for attention (e.g., '排名下滑 15 名')")


class PeriodInfo(BaseModel):
    """Current period metadata"""

    model_config = ConfigDict(from_attributes=True)

    period_id: str = Field(..., description="Period UUID as string")
    period_number: int = Field(..., ge=1)
    period_label: str = Field(..., description="Display label")
    start_date: str = Field(..., description="ISO date string")
    end_date: str = Field(..., description="ISO date string")
    days: int = Field(..., ge=1)


class AllianceAnalyticsResponse(BaseModel):
    """Complete alliance analytics data for AllianceAnalytics page"""

    model_config = ConfigDict(from_attributes=True)

    # Overview KPIs
    summary: AllianceSummary = Field(..., description="Alliance-wide metrics summary")

    # Trend across all periods (with medians)
    trends: list[AllianceTrendWithMedian] = Field(..., description="Period-by-period trend data")

    # Distribution histogram bins
    distributions: DistributionData = Field(..., description="Contribution and merit distributions")

    # Groups with box plot stats
    groups: list[GroupStatsWithBoxPlot] = Field(..., description="All groups with box plot data")

    # Top performers (top 10 by contribution)
    top_performers: list[PerformerItem] = Field(..., description="Top 10 performers by contribution")

    # Bottom performers (bottom 5 by contribution)
    bottom_performers: list[PerformerItem] = Field(..., description="Bottom 5 performers by contribution")

    # Needs attention members
    needs_attention: list[AttentionItem] = Field(..., description="Members needing attention")

    # Current period metadata
    current_period: PeriodInfo = Field(..., description="Latest period info")
